### ISSUE TITLE
Clausification of nested Implication fails

### ISSUE THREAD

**Issue by maweki** [pysathq/pysat#158]
I am not sure if I have some big misunderstanding, but I can't get clausification of conjunctions of implications to work. Conjunctions of other conjunctions or disjunctions I got to work fine.

Here is my minimal test case:

```python
import pysat
print(pysat.VERSION)
from pysat.formula import Atom, Implies, And
And(Implies(Atom(1), Atom(2))).clausify()
```

And the result is 

```
(1, 8, 'dev', 1)
Traceback (most recent call last):
  File "/home/maweki/s/python/orange-boxes/case.py", line 4, in <module>
    And(Implies(Atom(1), Atom(2))).clausify()
  File "/home/maweki/.local/lib/python3.12/site-packages/pysat/formula.py", line 1167, in clausify
    self._clausify(name_required=False)
  File "/home/maweki/.local/lib/python3.12/site-packages/pysat/formula.py", line 1594, in _clausify
    sub._clausify(name_required=True)
  File "/home/maweki/.local/lib/python3.12/site-packages/pysat/formula.py", line 2134, in _clausify
    self.clauses.append([self.name, -self.clauses[0][i]])
                                     ~~~~~~~~~~~~~~~^^^
IndexError: list index out of range
```

Clausify of the implication itself works fine.

But using `And`, `Or`, `Neg` to nest the implication leads to this error.

```python
Or(Implies(Atom(1), Atom(2))).clausify()
And(Implies(Atom(1), Atom(2))).clausify()
Neg(Implies(Atom(1), Atom(2))).clausify()
```

All of these fail

Edit:
it seems to me, that this codepath must raise an exception as it always goes out of bounds

https://github.com/pysathq/pysat/blob/a3266647552e72abdab838944817e2f92a06fc00/pysat/formula.py#L2132C1-L2134C70

```python
            # clauses representing converse implication
            for i in range(1, len(self.clauses[0]) + 1):
                self.clauses.append([self.name, -self.clauses[0][i]])
```

**Comment by alexeyignatiev** [comment]
Well, `And` should expect at least two arguments in the first place, @maweki. What are you "conjoining" with a single argument? I guess the implementation should throw an exception when that happens - it does not and I will need to fix this. Irregardless, this is certainly a bug requiring a fix. I will look into this when I get a chance. Thanks for reporting.

**Comment by maweki** [comment]
> Well, And should expect at least two arguments

No, it should not. A conjunction with one argument is not a special case. It is the identity. And throwing an exception would need a lot of error handling on the user side.

Also, the documentation is clear that any (natural) number of arguments is allowed (https://pysathq.github.io/docs/html/api/formula.html#pysat.formula.And). i.e., a conjunction...

Anyways, this is just the minimal example. My conjunctions are usually longer.

**Comment by alexeyignatiev** [comment]
@maweki, I've fixed this now. Hopefully, it should work. If you had a chance to play with the rest of the new formula functionality and reported bugs, it would be really helpful! 🙂

**Comment by alexeyignatiev** [comment]
> > Well, And should expect at least two arguments
> 
> No, it should not. A conjunction with one argument is not a special case. It is the identity. And throwing an exception would need a lot of error handling on the user side.

The current implementation of the clausifier won't handle this. So it's either left to the user to make it a proper conjunction/disjunction/equivalence, or the clausifier needs to get more complicated to handle special cases.

**Comment by maweki** [comment]
> The current implementation of the clausifier won't handle this.

That's not true. One argument for the conjunction and disjunction works fine and reduces to the identity.

**Comment by alexeyignatiev** [comment]
OK, conjunctions and disjunctions do handle single arguments properly while equivalence and exclusive disjunctions fail to do so, which is to be expected. I am reverting the change for `And` and `Or`, respectively, and will upload the new version shortly.

**Comment by maweki** [comment]
Thank you very much. Maybe `And(*[])` and `Or(*[])` (as I use it) could be put on the roadmap and reduce to `True` and `False` respectively.

**Comment by alexeyignatiev** [comment]
I agree it would make sense to do at some point! 🙂

**Comment by maweki** [comment]
So I think it is most easily added in the following way:

replace

```python
    def __new__(cls, *args, **kwargs):
        """
            Conjunction constructor.
        """

        return Formula.__new__(cls, args, kwargs, type=FormulaType.AND)
```

with

```python
    def __new__(cls, *args, **kwargs):
        """
            Conjunction constructor.
        """
        if args:
            return Formula.__new__(cls, args, kwargs, type=FormulaType.AND)
        return PYSAT_TRUE
```

and for the disjunction

```python
    def __new__(cls, *args, **kwargs):
        """
            Disjunction constructor.
        """

        if args:
            return Formula.__new__(cls, args, kwargs, type=FormulaType.OR)
        return PYSAT_FALSE
```

If you're happy with that, I'll do a PR.

**Comment by maweki** [comment]
It wouldn't. Okay. I'll take another look.

**Comment by maweki** [comment]
Okay. For conjunction, we only need to remove this check here:

```python
        if len(self.subformulas) < 1:
            raise FormulaError('Conjunction requires at least 1 argument')
```

As the for-loop doesn't happen and `operands` stays empty, `simplify` works out fine and reduces to True

https://github.com/pysathq/pysat/blob/5c2ade46c70e0e028cca4562a65284f4a603079c/pysat/formula.py#L1550C1-L1567C30

```python
        for sub in self.subformulas:
            # execute simplification recursively
            sub = sub.simplified(assumptions=set(assumptions))

            # simplify the current term
            if sub is PYSAT_FALSE:
                return PYSAT_FALSE

            # a new yet unsimplified term
            elif sub != PYSAT_TRUE and sub not in oset:
                oset.add(sub)
                operands.append(sub)

        # there is nothing left and we did not come across False
        # we should return True
        if not operands:
            return PYSAT_TRUE
```

Clausification correctly does nothing with the empty list of subformulas and there are no clauses. Maybe `__str__` should be `str(Atom(True))` in this case. Not sure what's nicest.

The same is true for the disjunction. Just removing the check should be fine, if we also make a better print statement. Clausification creates an empty clause, that represents false. Correct. Simplification reduces to False.

As far as I can see, everything already works out nicely, which does not surprise me. The papers you reference don't handle this as a special case either ;)

**Comment by alexeyignatiev** [comment]
Hm, this does not seem to work due to problems in argument flattening. I will have to look into this more carefully when addressing the corresponding bug [you raised](https://github.com/pysathq/pysat/issues/160).

**Comment by alexeyignatiev** [comment]
This should be addressed in [1.8.dev7](https://github.com/pysathq/pysat/commit/906d9e305428ba404564cd5c444e6d5a5128716b). Closing.
